### ISSUE TITLE
Make the tox API explicit

### ISSUE THREAD

**Issue by obestwalter** [tox-dev/tox#800]
tox has been growing some kind of programmatic API over the years, without it being properly acknowledged or any kind of deprecation handling. Make the current API explicit and introduce a simple deprecation strategy to give users some guidance.

**Comment by obestwalter** [comment]
While spring cleaning I got a bit carried away and though this was a no brainer from what I learned since being more involved in OSS projects, but I would really appreciate some feedback especially from some more experienced OSS maintainers.

My thinking was that I constantly observe projects grow implicit de-facto APIs - like e.g. the "pip API" that never existed but AFAICT still was in common use (programmatic use of pip was always actively discouraged). Same with pytest internals, although it having a very exhaustive plugin system. In a language like Python where every internal of every library you use can be accessed, monkeypatched, deleted, etc. this happens and it is o.k. - we're all consenting adults after all. Users doing this should be aware though that these internals can change without warning - even if they are not explicitly marked as private (`_` at beginning of the name for private or `__` for **really** private). 

Introducing an explicit API is acknowledging this and for me a way of communicating where we try to keep it stable and give the user enough warning before breaking it (e.g. emit deprecation warnings for at least one minor release before breaking things).

tox has been quite stable over the last years, so this seems to be a good time for me to do this and I think it is also a good idea, because things might change more in the future to keep tox up to date with changes in packaging and new tools pooping up. A clear deprecation strategy for an "official" might cause less pain and friction then instead of breaking things without warning just because tox has no API.

So here it is for now:

https://github.com/tox-dev/tox/blob/f94da07f4a0db28432ecfd0d7ca5d29b2b43ed44/tox/__init__.py#L17-L28

The very minimum are `cmdline` (for IDEs and tools building on top of tox) and the import position of `hookimpl` (for plugins). About the rest I am not so sure, but it was all stuff that hung around in the root tox namespace and therefore invited itself to be imported.

`hookspec` was also explicitly made available in the root namespace, but I honestly don't know why, because it has no use for anyone outside of tox, unless I completely have something wrong there. This is why this is deprecated and marked for removal in tox 4.

If introducing the constants as part of the API is a good idea I honestly don't know, but at lease `default_factors` was already being used and I think that some of tox "knowledge" should actually be accessible for tools building on top of tox.

I also got aware that there needs to be some more documentation around the deprection stuff than there is already, but I think this can be taken pretty much unchanged from what we introduced for pytest a while ago.
